File: packages/python/secrefs/parser.py
"""
URI parser for SecRefs' `sec://` reference format:

    sec://<provider-alias>/<secret-path-or-id>[#<json-field>]

    sec://aws/prod/db#password
    sec://vault/secret/data/stripe#key
    sec://local/mock-db#password

Mirrors packages/node/src/parser.ts exactly (same regex, same semantics),
so a `sec://` string means the same thing regardless of which SDK reads it.
"""

from __future__ import annotations

import re
from dataclasses import dataclass
from typing import Optional

_SEC_REF_PATTERN = re.compile(r"^sec://([a-zA-Z0-9][a-zA-Z0-9_-]*)/([^\s#]+)(?:#([^\s#]+))?$")


class SecRefParseError(ValueError):
    def __init__(self, raw: str, reason: str) -> None:
        self.raw = raw
        self.reason = reason
        super().__init__(f'Invalid secret reference "{raw}": {reason}')


@dataclass(frozen=True)
class ParsedSecretRef:
    raw: str
    provider: str
    path: str
    field: Optional[str] = None


def is_secret_ref(value: object) -> bool:
    """True if `value` is a string that looks like a `sec://` reference at all."""
    return isinstance(value, str) and value.startswith("sec://")


def parse_secret_ref(raw: object) -> ParsedSecretRef:
    """Parses a `sec://` reference string, raising SecRefParseError on any malformed input."""
    if not isinstance(raw, str):
        raise SecRefParseError(str(raw), "reference must be a string")

    trimmed = raw.strip()
    if not trimmed.startswith("sec://"):
        raise SecRefParseError(raw, 'must start with "sec://"')

    match = _SEC_REF_PATTERN.match(trimmed)
    if not match:
        raise SecRefParseError(raw, "does not match sec://<provider>/<path>[#field] format")

    provider, path, field = match.groups()
    if not provider:
        raise SecRefParseError(raw, "missing provider alias")
    if not path:
        raise SecRefParseError(raw, "missing secret path")

    return ParsedSecretRef(raw=raw, provider=provider.lower(), path=path, field=field or None)


def try_parse_secret_ref(raw: object) -> Optional[ParsedSecretRef]:
    """Best-effort parse that returns None instead of raising."""
    try:
        return parse_secret_ref(raw)
    except SecRefParseError:
        return None

File: packages/node/src/parser.ts
/**
 * URI parser for SecRefs' `sec://` reference format:
 *
 *   sec://<provider-alias>/<secret-path-or-id>[#<json-field>]
 *
 *   sec://aws/prod/db#password
 *   sec://vault/secret/data/stripe#key
 *   sec://local/mock-db#password
 *
 * The provider alias is a bare identifier (letters/digits/`-`/`_`), the path
 * is opaque to this parser (providers interpret it however their backend
 * needs), and the optional `#field` fragment supports dot-notation for
 * traversing nested JSON secrets (e.g. `#nested.value`).
 */

const SEC_REF_PATTERN = /^sec:\/\/([a-zA-Z0-9][a-zA-Z0-9_-]*)\/([^\s#]+)(?:#([^\s#]+))?$/;

export interface ParsedSecretRef {
  /** The original, unmodified reference string. */
  raw: string;
  /** Lowercased provider alias, e.g. "aws", "vault", "local". */
  provider: string;
  /** The secret path/id as understood by the provider. */
  path: string;
  /** Optional dot-notation field to extract from a JSON secret. */
  field?: string;
}

export class SecRefParseError extends Error {
  constructor(
    public readonly raw: string,
    public readonly reason: string,
  ) {
    super(`Invalid secret reference "${raw}": ${reason}`);
    this.name = "SecRefParseError";
  }
}

/** True if `value` is a string that looks like a `sec://` reference at all. */
export function isSecretRef(value: unknown): value is string {
  return typeof value === "string" && value.startsWith("sec://");
}

/**
 * Parses a `sec://` reference string. Throws {@link SecRefParseError} if the
 * value isn't a string, doesn't start with `sec://`, or doesn't match the
 * full `<provider>/<path>[#field]` shape.
 */
export function parseSecretRef(raw: unknown): ParsedSecretRef {
  if (typeof raw !== "string") {
    throw new SecRefParseError(String(raw), "reference must be a string");
  }

  const trimmed = raw.trim();
  if (!trimmed.startsWith("sec://")) {
    throw new SecRefParseError(raw, 'must start with "sec://"');
  }

  const match = SEC_REF_PATTERN.exec(trimmed);
  if (!match) {
    throw new SecRefParseError(
      raw,
      "does not match sec://<provider>/<path>[#field] format",
    );
  }

  const [, provider, path, field] = match;
  if (!provider) {
    throw new SecRefParseError(raw, "missing provider alias");
  }
  if (!path) {
    throw new SecRefParseError(raw, "missing secret path");
  }

  return {
    raw,
    provider: provider.toLowerCase(),
    path,
    field: field || undefined,
  };
}

/** Best-effort parse that returns `null` instead of throwing. */
export function tryParseSecretRef(raw: unknown): ParsedSecretRef | null {
  try {
    return parseSecretRef(raw);
  } catch {
    return null;
  }
}
